install-usercss/install-usercss.js

```diff
@@ -391,47 +317,80 @@
     }
   }
 
-  function getCodeDirectly() {
-    // FF applies page CSP even to content scripts, https://bugzil.la/1267027
-    // To circumvent that, the bg process downloads the code directly
-    const key = 'tempUsercssCode' + tabId;
-    chrome.storage.local.get(key, data => {
-      const code = data && data[key];
-
-      // bg already downloaded the code
-      if (typeof code === 'string') {
+  function initLiveReload() {
+    const DELAY = 500;
+    let isEnabled = false;
+    let timer = 0;
+    let sequence = Promise.resolve();
+    if (tabId >= 0) {
+      filePort = chrome.tabs.connect(tabId, {name: 'downloadSelf'});
+      filePort.postMessage('init');
+      filePort.onMessage.addListener(onPortMessage);
+      filePort.onDisconnect.addListener(onPortDisconnect);
+    }
+    return {
+      get enabled() {
+        return isEnabled;
+      },
+      onToggled(e) {
+        if (e) isEnabled = e.target.checked;
+        if (installed || installedDup) {
+          (isEnabled ? start : stop)();
+          $('.install').disabled = isEnabled;
+          $('#live-reload-install-hint').hidden = !isEnabled;
+          $('#live-reload-install-hint-ff').hidden = !isEnabled || !filePort;
+        }
+      },
+    };
+    function onPortMessage({code, error}) {
+      if (error) {
+        messageBox.alert(error, 'pre');
+      } else if (!initialized) {
         initSourceCode(code);
-        chrome.storage.local.remove(key);
-        return;
-      }
-
-      // bg still downloads the code
-      if (code && code.loading) {
-        const waitForCodeInStorage = (changes, area) => {
-          if (area === 'local' && key in changes) {
-            initSourceCode(changes[key].newValue);
-            chrome.storage.onChanged.removeListener(waitForCodeInStorage);
-            chrome.storage.local.remove(key);
-          }
-        };
-        chrome.storage.onChanged.addListener(waitForCodeInStorage);
-        return;
+      } else {
+        update(code);
       }
-
-      // on the off-chance dbExecChromeStorage.getAll ran right after bg download was saved
-      download(params.get('updateUrl'))
-        .then(initSourceCode)
-        .catch(err => messageBox.alert(t('styleInstallFailed', String(err)), 'pre'));
-    });
-  }
-
-  function onPortDisconnected() {
-    chrome.tabs.get(tabId, tab => {
-      if (chrome.runtime.lastError) {
-        closeCurrentTab();
-      } else if (tab.url === tabUrl) {
-        location.reload();
+    }
+    function onPortDisconnect() {
+      chrome.tabs.get(tabId, tab => {
+        if (chrome.runtime.lastError) {
+          closeCurrentTab();
+        } else if (tab.url === initialUrl) {
+          location.reload();
+        }
```

Comment:
Should we close the installer when the URL doesn't match?

Reply:
I thought we didn't previously. I have no opinion on this particular thing but generally I prefer not to change behavior out of the blue.